src/qat/purr/compiler/builders.py

```diff
@@ -553,89 +596,111 @@ def measure(
-            final_syncronize = syncs[0] + qubit
-            self._instructions.remove(syncs[0])
+        if full_measure_block and len(pre_syncs) >= 2 and len(pre_phase_resets) >= 1:
+            # Merge the first and last sync with the new.
+            pre_syncs[0] += new_syncs[0]
+            pre_syncs[-1] += new_syncs[-1]
+
             # reset all qubits
-            final_phase_reset = phase_resets[0] + entangled_qubits
-            self._instructions.remove(phase_resets[0])
+            pre_phase_resets[-1] += new_phase_resets[-1]
 
             # Add in our current changes.
-            self.add(
+            self.insert(
                 [
-                    MeasurePulse(measure_channel, **qubit.pulse_measure),
-                    acquire_instruction,
-                    final_syncronize,
-                    final_phase_reset,
-                ]
+                    val
+                    for val in new_measure_block
+                    if isinstance(val, (MeasurePulse, Acquire))
+                ],
+                index=self._instructions.index(pre_syncs[-1]),
             )
+        else:
+            self.add(new_measure_block)
+
+    def measure(
+        self,
+        qubit: Qubit,
+        axis: ProcessAxis = None,
+        output_variable: str = None,
+    ) -> "QuantumInstructionBuilder":
+        """
+        Adds a measure instruction. Important note: this only adds the instruction, not
+        any post-processing instructions as well. If you're wanting to perform generic
+        common operations use the more specific measurement types, as they add all the
+        additional information.
+        """
+        if axis is None:
+            axis = ProcessAxis.SEQUENCE
 
-            # Move all post-processing until after our newly-shifted syncronize block.
-            # Order matters here.
-            for pp in [
-                val
-                for val in reversed(previous_measure_block)
-                if isinstance(val, PostProcessing)
-            ]:
-                self._instructions.remove(pp)
-                self.add(pp)
+        if axis == ProcessAxis.SEQUENCE:
+            mode = AcquireMode.INTEGRATOR
+        elif axis == ProcessAxis.TIME:
+            mode = AcquireMode.SCOPE
         else:
-            self.add(
-                [
-                    Synchronize(entangled_qubits),
-                    MeasurePulse(measure_channel, **qubit.pulse_measure),
-                    acquire_instruction,
-                    Synchronize(qubit),
-                    PhaseReset(entangled_qubits),
-                ]
+            raise ValueError(f"Wrong measure axis '{str(axis)}'!")
+
+        entangled_qubits = list(self._entanglement_map.get(qubit, []))
+        previous_measure_block = self._find_previous_measurement_block()
+
+        if isinstance(previous_measure_block, list):
+            new_measure_block, acquire_instruction = self._generate_legacy_measure_block(
```

Comment:
Nice back combability! Is this required? I can't think of many cases where we need to merge old measure list people have manually inputted

Reply:
This is to make current usages still work as before. There will be a warning message added to this so that users are notified that this is no longer the standard operation.

Reply:
There's only a few places this is used used like this in Maac that properly need refactoring